### ISSUE TITLE
Introduce new config options to Field definition (auth, pre/post resolve, cache?), etc

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#281]
So, in GraphQL when fields are defined we have a `name`, `description`, `type`, and `resolve`. . .but nothing is stopping us from having additional data as part of the field registration that we can make use of. 

We could have fields registered something like this: 

```
$fields = [
   'myPrivateField' => [
       'type' => Types::String(),
       'description' => __( 'The description for my private field', 'wp-graphql' ),
       'auth' => [
           'isPrivate' => true, 
           'caps' => [ 'manage_options' ], // If set, this forces authentication **AND** specific authorization based on capability
       ],
       'preResolve' => $callable,
       'postResolve' => $callable,
       'shouldPrimeMetaCache' => true,
       'shouldPrimeTermCache' => true,
      'storageType' => 'post_meta',
   ],
];
```

**auth**
Auth could be a nested array with additional settings. My initial thoughts are we could have:

- isPrivate: If true, this would enforce authentication (any role/caps allowed, just need to be authenticated)
- caps: An array of capabilities to check to ensure the user has proper authorization to resolve the field. If this is set, it automatically overrides `isPrivate` as any capability check requires authentication first.

**preResolve / postResolve**
These would provide a spot for callbacks to be centralized around resolvers. We could automatically fire off a `graphql_before_resolve` action and a `graphql_after_resolve` action (don't exist currently) and this would provide a central place to define functionality for what should happen at those hooks for specific fields. 

You could of course _still_ do something to this tune outside of the Schema definition though too: 

```
add_action( 'graphql_after_resolve', function( $field ) { 
    if ( 'myCustomField' === $field['name'] ) {
        emailKevin();
    }
} );
```

**shouldPrimeMetaCache**
This could be used by lookAhead queries to determine if `update_post_meta_cache` should be applied to the WP_Query (default would be true, and the core post_type fields would be set to false so that any new fields would prime the cache unless they opt out)

**shouldPrimeTermCache**
This could be used by lookAhead queries to determine if `update_term_meta_cache` should be applied to the WP_Query (default would be true, and the core post_type fields would be set to false so that any new fields would prime the cache unless they opt out)

**storageType**
This one feels maybe a little bit more plugin territory. . .

Options could be `post_meta`, `term_meta`, `external`, etc. . .and this could be potentially used to do interesting things elsewhere in code such as generate a static list of all fields that are stored in a certain place...like `get_graphql_post_meta_fields()` or `get_graphql_external_fields()` or something to that tune, that could in turn be used to do something like filter all the fields in that group at once, or something along those lines. . .not sure, but feels useful. . .

I think DFM Transients has something along the lines of defining where a cache is stored, so there might be stuff to look at there for comparing? (or maybe not. . .)

**Comment by CodeProKid** [comment]
oOoOoOo, I like `isPrivate` for requiring auth.

**Comment by jasonbahl** [comment]
Initial implementation of this (auth) has been addressed by #291. 

I’m going to leave this open for further discussion, but boot it out of the 1.0 milestone as the _main_ piece of this is now in place so it’s just deciding what else should be part of the core plugin field definition options

**Comment by jasonbahl** [comment]
Closing as the initial implementation is in place. Further implementation, if needed, can be opened in a more specific issue for it's case.
